pkg/api/validation/validation.go

```diff
@@ -1237,6 +1256,21 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
 			allErrs = append(allErrs, validateScaleIOVolumeSource(pv.Spec.ScaleIO, specPath.Child("scaleIO"))...)
 		}
 	}
+	if pv.Spec.LocalVolume != nil {
+		if numVolumes > 0 {
+			allErrs = append(allErrs, field.Forbidden(specPath.Child("localVolume"), "may not specify more than 1 volume type"))
+		} else {
+			numVolumes++
+			if !utilfeature.DefaultFeatureGate.Enabled(features.PersistentLocalVolumes) {
+				allErrs = append(allErrs, field.Forbidden(specPath.Child("localVolume"), "LocalVolumes are disabled by feature-gate"))
+			}
+			allErrs = append(allErrs, validateLocalVolumeSource(pv.Spec.LocalVolume, specPath.Child("localVolume"))...)
+			// StorageClass is required
+			if len(pv.Spec.StorageClassName) == 0 {
```

Comment:
* The PV/PVC controller will not bind PVC's that request a storageclass with a topologyKey specified.  This is why if we go with this approach, I think storageclass is required.  Otherwise, we do not know when to trigger this different logic for local volumes.  The alternative is, like you mentioned, another field in the PVC.  But I do not like this alternative, because I think that this is a detail that the user should not need to know.
* The discovery daemon has to be configured to create PVs with a StorageClass name.  I am thinking of having some sort of configmap, with directory -> storageclass mapping.  ie all the volumes under dir1 will be part of storageclass1, everything under dir2 will use storageclass2, etc.
* I think it is required in order to delay the PVC binding until a pod comes.  Maybe for alpha, we could survive without this behavior if you start all your local storage pods first.